### ISSUE TITLE
Access Client Hello before accepting connection

### ISSUE THREAD

**Issue by FlorianUekermann** [quinn-rs/quinn#2024]
I need to access the `rustls::server::ClientHello` or at least the content of `server_name()` before accepting a connection with a rustls config. `alpn()` would also be very useful. Could this be exposed via `Incoming` or is that too early?

**Comment by Ralith** [comment]
`Incoming` is too early: it is yielded before the endpoint even decrypts first packet, and a `ClientHello` might span multiple packets. I think your use case is worth supporting, but it will require an additional intermediate stage between `Incoming` and `Connection`, and changes in the proto layer to suspend work once the `ClientHello` is received.

**Comment by FlorianUekermann** [comment]
I see. Sounds like a non-trivial amount if work. That's unfortunate. Maybe it would be easier to add an accept method, which takes a closure mapping Client Hello to a rustls config as a first step.

**Comment by djc** [comment]
We already have a `Connecting` stage, so this might fit in there? I wonder if we could do this by adding a (provided) method to quinn-proto's `ServerConfig` trait.

**Comment by FlorianUekermann** [comment]
> We already have a `Connecting` stage, so this might fit in there?

`Connecting` exposes the rustls handshake data (including the negotiated protocol), so I don't think we can get to hat stage without a rustls config. So even changing the semantics of the `Connecting` stage slightly would be tricky for compatibility reasons.

But the fact that the handshake data is available in `Connecting`, which you get by calling `Incoming::accept()` (which is not `async`, so presumably not waiting for network traffic) is a little hard to reconcile with @Ralith statement. Doesn't that mean the `Client Hello` would need to be received before we reach the `Incoming` stage? I'll try to take a look.

**Comment by Ralith** [comment]
`Connection::handshake_data` is async. If your Client Hello has not been fully received, it won't be ready yet.

**Comment by Ralith** [comment]
> We already have a Connecting stage, so this might fit in there?

I'm not sure if this would be feasible. The motivation here is to make the connection's configuration depending on the Client Hello, so we have to defer the connection logic until after the application has had its opportunity to weigh in, similar to how `Incoming` defers all frame processing.

**Comment by FlorianUekermann** [comment]
Adding an async `Incoming::client_hello_data`, which waits until the data is available similar to `Connecting::handshake_data` would be best from a user perspective imo. I don't think the public API would benefit from an intermediate stage, because I think it would just duplicate the methods `Incoming` already has.

**Comment by Ralith** [comment]
That could work. I think it'd need about the same changes at the proto layer: we still want decode a Client Hello but defer all other work, and unless the user calls `client_hello_data` we shouldn't even do that much. I agree that adding more stages with duplicate APIs is unappealing if it can be avoided, though there might be other viable approaches as well, like a single type parameterized by a stage marker type.

**Comment by FlorianUekermann** [comment]
Stupid question: Reaching `Incoming` means the "initial" packet (type `0x00`) has been received, right? Doesn't that mean the first CRYPTO frame, which includes the Client Hello has already been received as well?

**Comment by djc** [comment]
> Stupid question: Reaching `Incoming` means the "initial" packet (type `0x00`) has been received, right? Doesn't that mean the first CRYPTO frame, which includes the Client Hello has already been received as well?

Yes, the first CRYPTO frame is included, but there's no guarantee the complete ClientHello has been received (particularly with post-quantum key exchange).

**Comment by taikulawo** [comment]
Does quinn accept API breaking change? We are deeply rely on lazy select server_config.

Could we accept callback style like [s2n](https://github.com/aws/s2n-quic/blob/main/examples/async-client-hello-callback/src/bin/quic_async_client_hello_callback_server.rs) or LazyAcceptor like rustls?

**Comment by djc** [comment]
> Could we accept callback style like [s2n](https://github.com/aws/s2n-quic/blob/main/examples/async-client-hello-callback/src/bin/quic_async_client_hello_callback_server.rs) or LazyAcceptor like rustls?

I think we'd probably prefer something closer to rustls, mainly so callers can asynchronously decide which `ServerConfig` they want to use.

**Comment by taikulawo** [comment]
Our company have rust based gateway, try to add quic/http3 on it. 

So I am researching a Rust QUIC library(quinn/msquic/s2n/quiche/quictls) to integrate it into the gateway.

**Comment by taikulawo** [comment]
You may know this issue. I solved by patch rustls. https://github.com/rustls/rustls/issues/1976. since it cause breaking api change. I haven't submit PR.

I prefer use quic library which can integrate rustls as tls backend, so we can reuse previous patched rustls.

**Comment by djc** [comment]
Quinn integrates rustls already, and also makes the encryption layer pluggable so you could use your own rustls fork instead.

**Comment by taikulawo** [comment]
> Quinn integrates rustls already, and also makes the encryption layer pluggable so you could use your own rustls fork instead.

Yes, but quinn don't support select ServerConfig. In Tcp tls, we have TlsClientParse, We pre-parse tcp data get SNI before rustls get involved, so we can select different ServerConfig. But quic encapsulate TLS into itself, we can't pre-parse simple

**Comment by djc** [comment]
> > Quinn integrates rustls already, and also makes the encryption layer pluggable so you could use your own rustls fork instead.
> 
> Yes, but quinn don't support select ServerConfig. In Tcp tls, we have TlsClientParse, We pre-parse tcp data get SNI before rustls get involved, so we can select different ServerConfig. But quic encapsulate TLS into itself, we can't pre-parse simple
> 
> You can think we are NGINX, we have many domain listen on same quic port, we need use different cert/pkey. NGINX use openssl SSL_CTX_set_tlsext_servername_callback callback, so we need find rustls style way to solve it. :)

I think that's definitely feasible, someone just needs to do the work (or someone can pay me to do the work, happy to discuss).

**Comment by taikulawo** [comment]
> Our company have rust based gateway, try to add quic/http3 on it.
> 
> So I am researching a Rust QUIC library(quinn/msquic/s2n/quiche/quictls) to integrate it into the gateway.

On the first gray version, we will have 1,000,000 qps http/3 traffic which power our core business. So quic library MUST stable and reliable.

**Comment by iadev09** [comment]
I put together a small working demo for this discussion:

https://github.com/iadev09/quinn-server-config-from-clienthello-demo

It uses a Quinn prototype branch here:

https://github.com/iadev09/quinn/tree/server-config-from-clienthello

This is one concrete prototype for the callback-style direction discussed above. It keeps the existing `Incoming` flow unchanged. Since the ClientHello may span multiple packets, the prototype waits until the complete ClientHello is available, selects an `Arc<rustls::ServerConfig>`, then continues with the normal rustls QUIC handshake.

This version is deliberately synchronous and aimed at in-memory/snapshot-based selection. It is not a passive "expose ClientHello data on `Incoming`" API; the goal is the narrower deferred configuration-selection boundary: parse enough ClientHello to choose the final `rustls::ServerConfig`, then continue the handshake. It also does not try to solve async configuration lookup.

The shape is roughly:

```rust
let crypto = ClientHelloServerConfig::new(default_server_config, move |hello: ClientHello<'_>| {
    match hello.server_name() {
        Some("localhost") => Some(selected_server_config.clone()),
        _ => None,
    }
})?;

let server_config = quinn::ServerConfig::with_crypto(Arc::new(crypto));
```

The selected value is a complete `Arc<rustls::ServerConfig>`, so it can support per-host TLS policy such as a different client-auth verifier, resumption boundary, ALPN policy, or host-wide early-data capability.

The demo keeps the runnable code small. It creates two server configs:

- a selected config for `localhost`, trusted by the client
- a default config for `default.example`, not trusted by the client

If the ClientHello-based selection does not happen before rustls continues the QUIC handshake, the client certificate verification fails. With the prototype branch, the demo succeeds.

- Existing `QuicServerConfig` usage is unchanged.
- The new behavior is opt-in.
- The cost is scoped to users of the wrapper: buffer ClientHello bytes, parse the small selection fields, construct the selected rustls QUIC server connection, then replay the buffered bytes into rustls.
- This does not try to solve a general rustls-level acceptor API; it only targets Quinn's rustls integration point.

The README has a more detailed breakdown of the demo, API changes, and tradeoffs:

https://github.com/iadev09/quinn-server-config-from-clienthello-demo#readme

I would be interested in whether this is a direction worth refining into an upstream PR.

**Comment by djc** [comment]
I would like to see a solution that is all in on actual async-consumable type similar to what exists in rustls and tokio-rustls. That really needs more work to also extract the use of the rustls `ServerConfig` from `Endpoint`s.

I don't think callback APIs like this are really a great solution for the long term. It might be okay as a temporary workaround on the 0.11.x branch to bridge the gap towards a full solution that requires semver-incompatible changes for 0.12.

**Comment by iadev09** [comment]
That makes sense, thanks.

I agree the callback shape is probably not the right long-term API.

For the next step, would you prefer that I focus only on a Quinn-side acceptor-style API for 0.12, or should I also keep a rustls-side QUIC `Acceptor` shape in scope as a follow-up?

A Quinn-only design does not strictly require rustls changes: Quinn can buffer enough handshake data and then continue with the selected rustls `ServerConfig`. The tradeoff is that Quinn has to parse enough ClientHello itself.

A rustls-side QUIC `Acceptor` would avoid that duplicated parsing and may be cleaner long term, but it would depend on rustls API work. If that direction is worth exploring, should it target the current Quinn rustls dependency line, or rustls main (0.24)?

If that matches your view, I can sketch the 0.12 Quinn API direction before writing more code.

**Comment by djc** [comment]
> A rustls-side QUIC `Acceptor` would avoid that duplicated parsing and may be cleaner long term, but it would depend on rustls API work. If that direction is worth exploring, should it target the current Quinn rustls dependency line, or rustls main (0.24)?

I think this is probably the right direction, and I think it'd be okay to target this for rustls 0.24+ only.

Thanks for working on this!

**Comment by iadev09** [comment]
Small update from testing the rustls-0.24 acceptor direction:

While wiring this through `tokio-rustls::LazyConfigAcceptor`, I hit an existing TLS 1.3 continuation-path issue in rustls' server `Acceptor`: after `Accepted::into_connection()`, the selected protocol version is not propagated into the record-layer send/receive state, so TLS 1.3 compatibility `ChangeCipherSpec` records can be rejected.

Ironically, the TCP-side lazy acceptor path turned out to be a useful regression test while exploring the same acceptor-style shape for Quinn.

This is separate from the Quinn API shape, and I’m keeping it as a small rustls-side fix rather than mixing it into the Quinn work. The existing tokio-rustls lazy acceptor test reproduces it, and the fix is small.

I’ll continue treating the Quinn side as the 0.12 async acceptor-style API work, with the rustls-side QUIC `Acceptor` targeting rustls 0.24+ as discussed above.

**Comment by iadev09** [comment]
Small follow-up from trying the rustls-0.24 QUIC acceptor direction in a Quinn branch.

The branch has an in-repo Quinn regression test for the proposed boundary:

```text
incoming_acceptor_exposes_client_hello
```

The test accepts a connection through the staged path, verifies SNI/ALPN from rustls' ClientHello, then continues the handshake with a selected server config.

I run it with the rustls branch patched in:

```bash
cargo \
  --config 'patch.crates-io.rustls.git="https://github.com/iadev09/rustls.git"' \
  --config 'patch.crates-io.rustls.branch="quic-acceptor-tls13-ccs"' \
  --config 'patch.crates-io.rustls-aws-lc-rs.git="https://github.com/iadev09/rustls.git"' \
  --config 'patch.crates-io.rustls-aws-lc-rs.branch="quic-acceptor-tls13-ccs"' \
  --config 'patch.crates-io.rustls-ring.git="https://github.com/iadev09/rustls.git"' \
  --config 'patch.crates-io.rustls-ring.branch="quic-acceptor-tls13-ccs"' \
  --config 'patch.crates-io.rustls-util.git="https://github.com/iadev09/rustls.git"' \
  --config 'patch.crates-io.rustls-util.branch="quic-acceptor-tls13-ccs"' \
  test -p quinn incoming_acceptor_exposes_client_hello \
  --features runtime-tokio,rustls-aws-lc-rs \
  --no-default-features \
  --locked \
  -- --nocapture
```

The normal non-selection path stays unchanged:

```rust
while let Some(incoming) = endpoint.accept().await {
    tokio::spawn(async move {
        let connection = incoming.await?;
        handle_connection(connection).await;
        Ok::<_, quinn::ConnectionError>(())
    });
}
```

Callers that need ClientHello-based config selection can opt into the staged path:

```rust
let accepted = incoming.acceptor()?.await?;
let selected = select_server_config(accepted.client_hello()).await?;
let connection = accepted.accept(selected)?.await?;
```

In the current branch, `selected` is an `Arc<quinn::ServerConfig>` whose crypto config wraps the selected rustls server config. That was the smallest shape for the prototype because Quinn's existing accept path already works in terms of `quinn::ServerConfig`, and the continued accept path also reads transport-related settings from that type while constructing transport parameters.

I would not treat `Arc<quinn::ServerConfig>` as the proposed final API; it is just the current prototype's lowest-friction integration point. It may be cleaner for the final API to select only a rustls server config, or a rustls-backed crypto config, while keeping the surrounding Quinn transport/default endpoint configuration from the original server config. The part I am trying to validate here is the boundary itself: read ClientHello, choose the final server-side TLS/QUIC config, then continue the handshake.

This is a primitive API rather than a built-in `LazyConfigAcceptor`.[^lazy-note] A typed helper could be built above it if useful:

```rust
let acceptor = LazyH3Acceptor::new(selector);
let connection = acceptor.accept(incoming).await?;
```

I also tried the same shape in a downstream HTTP/3 server with a mixed public/mTLS listener; the wrapper keeps public hosts on the normal `Incoming` path and only uses the staged path for per-host TLS policy.

So, true confidence can still stand in the open.

Working branches:

https://github.com/iadev09/quinn/tree/server-config-acceptor

https://github.com/iadev09/rustls/tree/quic-acceptor-tls13-ccs

One open question for the API shape is how this should interact with future ECH handling. If SNI is encrypted, the selection boundary probably needs to stay aligned with rustls' view of the ClientHello/ECH result rather than assuming cleartext SNI in Quinn.

Does this look like the right direction to keep refining?

[^lazy-note]: While using `tokio-rustls::LazyConfigAcceptor` as the TCP/TLS comparison point on the rustls 0.24 line, I hit a small TLS 1.3 continuation-path issue in rustls' server `Acceptor`. I kept that as a separate rustls-side patch: https://github.com/iadev09/rustls/tree/fix-acceptor-tls13-ccs

**Comment by djc** [comment]
> Callers that need ClientHello-based config selection can opt into the staged path:
> 
> let accepted = incoming.acceptor()?.await?;
> let selected = select_server_config(accepted.client_hello()).await?;
> let connection = accepted.accept(selected)?.await?;
> 
> In the current branch, `selected` is an `Arc<quinn::ServerConfig>` whose crypto config wraps the selected rustls server config. That was the smallest shape for the prototype because Quinn's existing accept path already works in terms of `quinn::ServerConfig`, and the continued accept path also reads transport-related settings from that type while constructing transport parameters.

Sounds like a great direction!

**Comment by iadev09** [comment]
Thanks, that is helpful.

This direction seems to require related changes in both Quinn and rustls. It also needs a rustls-0.24-compatible Quinn base, plus a few temporary dependency pins, to keep the Quinn regression tests and downstream HTTP/3 tests runnable.

I do not currently see an upstream Quinn branch or PR preparing for rustls 0.24. To test this acceptor direction, I have had to keep a small stack of fork branches across Quinn and related crates:

- `rustls`: QUIC `Acceptor` prototype for 0.24+
- `quinn`: rustls-0.24 integration plus the staged `Incoming` acceptor API
- `tokio-rustls`, `hyper-rustls`, `rustls-platform-verifier`, `reqwest`: rustls-0.24 compatibility branches for runnable tests

However, the main purpose of this work is still the Quinn config-selection/acceptor boundary. The rustls-0.24 integration is working well enough to build and pass the Quinn regression test, but I have not tried to frame that dependency work as a separate upgrade proposal.

For now, I will keep refining this in my fork and exercise it in a controlled downstream deployment. That should give us more confidence in the API shape without forcing the review burden onto Quinn/rustls too early.

At the same time, I do think contributor review is important for this design. Different production gateway/server use cases may expose API constraints that my own integration testing would not catch, so a draft PR will probably become useful once the basic shape is ready for broader inline review.

My working plan is:

1. Keep the Quinn API focused on the staged primitive for now. If a higher-level helper is useful later, it should be layered on top rather than replace the exposed ClientHello boundary.
2. Continue testing it against the rustls QUIC `Acceptor` branch, since the two API shapes need to evolve together.
3. Add targeted stress and edge-case coverage around the staged path, mainly to find weak spots in the API shape rather than just prove that the prototype works.
4. Share findings here if they point to API changes, and wait for your guidance on when a draft PR would be useful for broader inline review.

**Comment by iadev09** [comment]
Follow-up on the rustls-0.24 QUIC acceptor direction.

I continued with the plan from my previous comment:

- keep the Quinn API focused on the staged primitive
- test it against the rustls QUIC `Acceptor` branch
- add targeted coverage around the staged path
- share findings here if they point to API-shape or implementation changes

At this point, the API shape I am testing is:

```rust
let accepted = incoming.acceptor()?.await?;

let connection = match selector(accepted.client_hello()) {
    Some(config) => accepted.accept_with(config)?.await?,
    None => accepted.accept()?.await?,
};
```

Here `selector` is application policy. Quinn exposes the staged ClientHello
boundary and lets callers either continue with the endpoint/default server
config or continue with a selected prebuilt `Arc<ServerConfig>`.

One important design constraint for me is that this should not turn into
"Quinn parses ClientHello and exposes a parallel view of it." The ClientHello
view should remain rustls' view. Quinn only needs enough staging to feed QUIC
TLS bytes into rustls, let rustls expose its `ClientHello`, and then continue
the QUIC connection with the chosen server config.

### Compatibility and cost model

The summary is: the existing single-config path stays unchanged; once callers
opt into the staged boundary, waiting for ClientHello is inherent; after fixing
pending Initial handling, the measured added cost is around noise/sub-10us in
local runs.

The three cases I am trying to keep separate are:

```text
single config:
  Incoming -> Connection
  unchanged

staged default:
  Incoming -> ClientHello -> selector(None) -> default ServerConfig -> Connection
  opt-in; cannot avoid waiting for ClientHello once this boundary is used

staged selected:
  Incoming -> ClientHello -> selector(Some(config)) -> selected ServerConfig -> Connection
  opt-in; current local measurements put the added cost around noise/sub-10us
  after the pending-Initial fix
```

One API-shape question: should Quinn stop at this staged primitive, or would
you prefer Quinn to own a small higher-level helper around it? My current bias
is still to keep the Quinn API primitive and let libraries/apps build their own
selector layer on top. This feels like an advanced/server-gateway use case
rather than the common single-config server path, so I am not sure Quinn needs
to own the higher-level policy abstraction itself.

The Quinn branch is here:

https://github.com/iadev09/quinn/tree/server-config-acceptor

The rustls branch it depends on is here:

https://github.com/iadev09/rustls/tree/quic-acceptor

This is the rustls-side piece I would expect to turn into a separate draft PR
in rustls. I am linking it here because the Quinn API shape depends on that
QUIC `Acceptor` boundary. It also lines up with rustls/rustls#1218, which
tracks QUIC support for `Acceptor`:

https://github.com/rustls/rustls/issues/1218

The branch adds a QUIC-compatible `server::Acceptor` path that can consume QUIC
TLS handshake bytes up to ClientHello, expose rustls' `ClientHello`, and then
continue the QUIC TLS session with a supplied `rustls::ServerConfig`. The
important boundary for Quinn is that ClientHello parsing stays inside rustls;
Quinn feeds QUIC crypto data into that staged acceptor and resumes the
connection after server config selection.

The external bench repo is here:

https://github.com/iadev09/quinn-acceptor-bench

Its README includes the benchmark results, reproduction commands, and the notes
for how I got from the initial slow staged prototype to the latest plan with
pending-Initial handling.

### Multi-Initial ClientHello

One important implementation issue showed up during benchmarking: with the default AWS-LC
provider, the hybrid `X25519MLKEM768` key share can make the ClientHello span
multiple Initial packets. My first staged prototype waited roughly 28-30 ms in
`incoming.acceptor()?.await?` because there was no full Quinn `Connection` yet
to ACK the first Initial packet while waiting for the rest of ClientHello.
This matters because post-quantum hybrid key exchange makes larger
ClientHellos a realistic default configuration, not just a synthetic edge case;
large deployments such as Cloudflare and Google/Chrome are already moving this
direction.

The larger default hybrid key share exposed a **valid multi-Initial
ClientHello** case that the staged path has to handle correctly.

The current branch fixes that by keeping a small pending Initial state while
the connection is staged:

- track received Initial packet numbers for ACK generation
- allocate and route a server local CID while still staged
- send ACK-only Initial packets when ClientHello is incomplete
- carry the consumed CRYPTO offset into the final `Connection`
- carry the next Initial packet number into the final `Connection`

That fix is intentionally Quinn-side transport work; rustls still only owns
parsing and continuing the QUIC TLS stream. It also only applies to the staged
acceptor path. The existing single-config path still constructs the full
`Connection` immediately, so the normal packet/frame/ACK machinery continues to
handle this case there.

After that, the staged paths are back in the same sub-millisecond range as the
single-config path in the local benchmark. The external bench repo has the
tables and commands, including a `--fresh-client` check to avoid relying on
client endpoint reuse. The remaining local difference should not be read as
selector or staged-API overhead. In this diagnostic setup it is mostly the cost
of the valid larger ClientHello taking the multi-Initial path locally. That
packetization effect is not specific to Quinn's selector policy; in a real
network, it should be viewed against the normal packetization and RTT behavior
of a larger ClientHello.

```text
staged selected:               mean 527 us, p50 559 us, p95 773 us, p99 900 us
staged selected --x25519-only: mean 321 us, p50 314 us, p95 428 us, p99 448 us
```

So the local diagnostic delta here is roughly 200 us, and it is mainly a
large-ClientHello / multi-Initial-path signal. It is useful for explaining why
the staged path must handle pending Initial ACKs correctly, but it should not be
presented as the cost of config selection itself.

The external repo is still better for explaining the performance story because
it also records the pre-fix 28-30 ms failure mode and the packetization
diagnosis.

I also added in-repo Quinn benchmark cases for the three relevant server accept
paths:

```text
handshake_single_config
handshake_staged_default
handshake_staged_selected
```

I am treating those as regression/sanity benchmarks. The useful signal is that
the staged paths now track the single-config path closely enough that the
overhead is not showing up as a separate order of magnitude.

At this point the shape feels coherent to me:

- rustls owns ClientHello parsing and exposes the QUIC acceptor boundary
- Quinn owns the pending Initial packet/CID handling while staged
- applications own the server-config selection policy
- the existing single-config accept path remains unchanged

I can keep refining this in the fork for now. If you think the Quinn shape is
ready for broader inline review, I can turn the Quinn branch into a draft PR.
The rustls-side branch can be handled separately in the rustls repo, where the
QUIC `Acceptor` API belongs.

**Comment by djc** [comment]
Can you summarize this in your own words rather than posting verbose LLM slop?

An important aspect of the rustls design is that the `Acceptor` is a separate object in the sense that there is not just a sync callback that can respond to the `ClientHello`, such that callers can read from the `ClientHello`, then do potentially asynchronous work before responding to it (like reading the appropriate server certificate from Redis).

**Comment by iadev09** [comment]
I made the previous comment too long because i felt to report every step open about how i came to this solution.
This was my first to hit  PQ issue and it was  visible only in benches

### Summary 
Quinn gives the caller access to rustls' `ClientHello`, then application code can do sync or async work before continuing the handshake.


This API also can also be used:

```rust
let accepted = incoming.acceptor()?.await?;
let hello = accepted.client_hello();

// Application code decides this, synchronously or asynchronously.
let config = app_state.select_server_config(hello).await?;

let connection = accepted.accept_with(config)?.await?; 

```


This snippet is a usage pattern written on top of that primitive:
```rust

let connection = match selector(accepted.client_hello()) {
    Some(config) => accepted.accept_with(config)?.await?,
    None => accepted.accept()?.await?,
};
```

This removes the need to choose one of those models, or it can enable for a sturct shape to be merged into Quinn. 

the rustls-side piece is not a callback either. It is a `server::Acceptor` object that owns the partial QUIC TLS handshake state.

```rust

/// A QUIC server-side acceptor.
///
/// `Acceptor` allows callers to choose a [`ServerConfig`] after reading the
/// [`ClientHello`] of an incoming QUIC connection.
pub struct Acceptor {
    inner: Option<ConnectionCommon<ServerSide>>,
}

```

Shortly, caller will build config in boot time, swap the config resover maybe during relaod, but again can do  async lookup to select ServerConfig while handhake starting.  

the callback solution was  one week before and withoout rustls and doesnt depend 0.24, we had discussed that might be used in quin 0.11 as legacy path.

**Comment by iadev09** [comment]
My focus was the operator-side limit: certificate selection can be worked around in several ways, but one `ServerConfig` can only represent one TLS policy.

Sorry, I skipped over the certificate-selection case too quickly and interpreted your Redis example too narrowly as an async certificate-resolver case.

Now, I pushed one more commit to make the shape concrete:

https://github.com/iadev09/quinn/commit/5f772f50

It adds a `ServerConfigSelector` trait as a typed helper over the staged
acceptor primitive:

```rust
impl ServerConfigSelector for MySelector {
    type Error = MyError;

    async fn select_server_config<'a>(
        &'a self,
        hello: rustls::server::ClientHello<'a>,
    ) -> Result<Option<Arc<ServerConfig>>, Self::Error> {
        // caller-owned async policy
    }
}

let connecting = incoming.accept_with_selector(&selector).await?;
let connection = connecting.await?;
```

A more complete usage shape looks like this:

```rust
use std::{convert::Infallible, net::SocketAddr, sync::Arc};

use quinn::{Endpoint, ServerConfig, ServerConfigSelector};

struct HostSelector {
    host_b: Arc<ServerConfig>,
}

impl ServerConfigSelector for HostSelector {
    type Error = Infallible;

    async fn select_server_config<'a>(
        &'a self,
        hello: quinn::rustls::server::ClientHello<'a>,
    ) -> Result<Option<Arc<ServerConfig>>, Self::Error> {
        // Caller-owned/implemeneted async block
        match hello.server_name().map(|name| name.as_ref()) {
            Some("host-b.test") => Ok(Some(self.host_b.clone())),
            _ => Ok(None),
        }
    }
}

async fn run(
    addr: SocketAddr,
    host_a_config: ServerConfig,
    host_b_config: ServerConfig,
) -> anyhow::Result<()> {
    // The endpoint still has a normal/default config. If the selector returns
    // None, the handshake continues with this config.
    let endpoint = Endpoint::server(host_a_config, addr)?;

    let selector = Arc::new(HostSelector {
        host_b: Arc::new(host_b_config),
    });

    while let Some(incoming) = endpoint.accept().await {
        let selector = selector.clone();

        tokio::spawn(async move {
            let connecting = match incoming.accept_with_selector(selector.as_ref()).await {
                Ok(connecting) => connecting,
                Err(error) => {
                    tracing::debug!("failed to select server config: {error}");
                    return;
                }
            };

            let connection = match connecting.await {
                Ok(connection) => connection,
                Err(error) => {
                    tracing::debug!("handshake failed: {error}");
                    return;
                }
            };

            handle_connection(connection).await;
        });
    }

    Ok(())
}
```

This is not meant to replace the lower-level staged object. It is just one
possible typed layer above it.

The lower-level primitive remains available for callers that want full control:

```rust
let accepted = incoming.acceptor()?.await?;
let hello = accepted.client_hello();
let config = app_state.select_server_config(hello).await?;
let connection = accepted.accept_with(config)?.await?;
```

For the acceptor/config-selection tests, I ran:

```bash
cargo \
  --config 'patch.crates-io.rustls.git="https://github.com/iadev09/rustls.git"' \
  --config 'patch.crates-io.rustls.branch="quic-acceptor"' \
  --config 'patch.crates-io.rustls-aws-lc-rs.git="https://github.com/iadev09/rustls.git"' \
  --config 'patch.crates-io.rustls-aws-lc-rs.branch="quic-acceptor"' \
  --config 'patch.crates-io.rustls-ring.git="https://github.com/iadev09/rustls.git"' \
  --config 'patch.crates-io.rustls-ring.branch="quic-acceptor"' \
  --config 'patch.crates-io.rustls-util.git="https://github.com/iadev09/rustls.git"' \
  --config 'patch.crates-io.rustls-util.branch="quic-acceptor"' \
  test -p quinn incoming_acceptor \
  --features runtime-tokio,rustls-aws-lc-rs \
  --no-default-features \
  --locked \
  -- --nocapture
```


The `Quinn `shape feels more native because the staged ClientHello boundary is
part of the `Incoming` lifecycle, rather than an adapter wrapped around an I/O
stream as in the TCP `LazyConfigAcceptor` path. 


This issue has also mixed a few related but different tracks, diffrent versions , so I am trying to separate them to follow more clearly:

- Quinn 0.12 staged acceptor: this  issue #2024  continues that direction.
- rustls 0.24+ QUIC Acceptor: I opened a draft PR in rustls for the rustls-side:  rustls/rustls#3070 (MERGED to rustls's main)
- Quinn 0.11 callback bridge:  a separate Quinn PR for legacy:  quinn-rs/quinn/#2661.  (REMOVED - due to complexity) 
   an alternative https://crates.io/crates/quinn-deferred-config

**Comment by iadev09** [comment]
### Status

I cleaned up the branch after rustls/rustls#3070 was merged.

**Branch:** https://github.com/iadev09/quinn/tree/server-config-acceptor

Commits in the branch:

- [Port rustls integration to upstream 0.24 acceptor](https://github.com/iadev09/quinn/commit/5936900d) (alignment, not an aws-lc proposal)
- [Add rustls acceptor session start path](https://github.com/iadev09/quinn/commit/89333f50)
- [Add proto rustls ClientHello acceptor](https://github.com/iadev09/quinn/commit/090645a2)
- [Expose staged acceptor from Incoming](https://github.com/iadev09/quinn/commit/163c3bac)
- [Test staged rustls acceptor paths](https://github.com/iadev09/quinn/commit/54de3d19)
- [Add ServerConfigSelector helper](https://github.com/iadev09/quinn/commit/9d526c38)

**Bench/demo:** https://github.com/iadev09/quinn-acceptor-bench

It now compares staged default `ServerConfig` vs selected `ServerConfig` with `aws-lc` (PQ) and `ring`.

**Tests:** from the Quinn branch

```sh
cargo test -p quinn incoming_acceptor \
  --features runtime-tokio,rustls-aws-lc-rs \
  --no-default-features \
  --locked \
  -- --nocapture
```

**Comment by djc** [comment]
Do you want to submit a PR? It's fine to have rustls as a git dependency for now. Please make the commits as small as possible, separating refactoring from new code (but make sure each commit compiles and should generally be able to pass CI).

**Comment by iadev09** [comment]
Ok,  I will plan this with commit groups in one branch, including rustls-0.24  upgrade and alignments.  So, it will be easy to split/pick later.
